src/Nethermind/Nethermind.Network/TrustedNodes/TrustedNodesManager.cs

```diff
@@ -0,0 +1,178 @@
+// SPDX-FileCopyrightText: 2024 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net.Sockets;
+using System.Runtime.CompilerServices;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+using Nethermind.Config;
+using Nethermind.Core.Crypto;
+using Nethermind.Logging;
+using Nethermind.Stats.Model;
+
+namespace Nethermind.Network
+{
+    public class TrustedNodesManager : ITrustedNodesManager
+    {
+        private ConcurrentDictionary<PublicKey, NetworkNode> _nodes = new();
+        private readonly string _trustedNodesPath;
+        private readonly ILogger _logger;
+
+        public TrustedNodesManager(string trustedNodesPath, ILogManager logManager)
+        {
+            _trustedNodesPath = trustedNodesPath.GetApplicationResourcePath();
+            _logger = logManager.GetClassLogger();
+        }
+
+        public IEnumerable<NetworkNode> Nodes => _nodes.Values;
+
+        public async Task InitAsync()
+        {
+            if (!File.Exists(_trustedNodesPath))
+            {
+                if (_logger.IsDebug) _logger.Debug($"Trusted nodes file not found at: {_trustedNodesPath}");
+            }
+
+            var nodes = new ConcurrentDictionary<PublicKey, NetworkNode>();
+
+            await foreach (string line in File.ReadLinesAsync(_trustedNodesPath))
+            {
+                if (string.IsNullOrWhiteSpace(line))
+                {
+                    continue;
+                }
+                try
+                {
+                    NetworkNode node = new NetworkNode(line);
+                    nodes.TryAdd(node.NodeId, node);
+                }
+                catch (Exception ex) when (ex is ArgumentException or SocketException)
+                {
+                    if (_logger.IsError)
+                    {
+                        _logger.Error($"Failed to parse '{line}' as a trusted node.", ex);
+                    }
+                }
+            }
+
+            if (_logger.IsInfo)
+            {
+                _logger.Info($"Loaded {nodes.Count} trusted nodes from: {Path.GetFullPath(_trustedNodesPath)}");
+            }
+            if (_logger.IsDebug && nodes.Any())
+            {
+                _logger.Debug("Trusted nodes:\n" + string.Join(Environment.NewLine, nodes.Values.Select(n => n.ToString())));
+            }
+
+            _nodes = nodes;
+        }
+
+
+        // ---- INodeSource requirement: IAsyncEnumerable<Node> ----
+        // C# requires 'async' for IAsyncEnumerable yield. We'll add a small 'await Task.Yield()'
+        // to avoid the warning about "no awaits".
+        public async IAsyncEnumerable<Node> DiscoverNodes([EnumeratorCancellation] CancellationToken cancellationToken)
```

Comment:
Looks like this method will enumerate nodes on start and will ignore added via RPC until restart. What if we add a channel or so?

Reply:
Oh I see, you tried to use channels. Still seems like adding peers during run is what would be expected from UX perspective. Channel can be created in ctor, it may receive updates in `AddAsync`. Iterate it in `DiscoverNodes`

Reply:
Hm... so maybe channel made sense here and I misunderstood something.